README.md
<div align="center">
<img src="https://i.imgur.com/Cd6mBVN.png">
<hr>

Illegal Services is an application created to facilitate access to the site of illegal downloads: streaming, torrent, direct download, cracks, DDoS...
Indeed, illegal sites must constantly change their domain names to counter government attacks. It's very annoying to have to find them every time.
This is why Illegal Services will be your best friend. Especially since it only chooses a list of the best sites, and no scam site is included.

But that's not all, Illegal Services is also a powerful Multi-Tool allowing other functionalities such as DDoS attacks, Port Scanning and IP Lookup.

All this to say that if you like watching movies, crack applications or doing DDoS attacks this is the perfect application created to serve you.

It is compatible with Windows 10 and 11 (x86/x64).

[![Telegram](https://img.shields.io/badge/Telegram-Illegal%20Services-28a8e9?logo=telegram&labelColor=28a8e9)](https://t.me/illegal_services_forum)
[![GitHub latest version](https://img.shields.io/badge/GitHub-Download%20Latest%20Version-2b3137?logo=github&labelColor=2b3137)](https://github.com/Illegal-Services/Illegal_Services/releases/latest/download/IS.Setup.exe)
  
</div><hr>

<div align="center">

**⚠️ Join my Telegram to get the latest updates ! : https://t.me/illegal_services_forum ⚠️**

<hr></div>


ChangeLog.txt
KNOWN BUGS:
https://github.com/Illegal-Services/Illegal_Services/issues
SOCIAL MEDIA'S:
https://illegal-services.github.io/Illegal_Services/
https://t.me/illegal_services_forum
https://t.me/illegal_services
https://github.com/Illegal-Services/Illegal_Services

v6.1.9.4 - 25/08/2023
* FIXED a bug from "curl.exe"'s "--create-dirs" option not working on cyrillic computers. (thanks @LegalDudee (https://t.me/LegalDudee) for reporting the bug, and testings)

v6.1.9.3 - 10/07/2023
* FIXED a bug introduced in patch v6.1.9.2 - 09/07/2023

v6.1.9.2 - 09/07/2023
* FIXED a bug when "OPEN" choice in Illegal Services x IS.bookmarks.html where showing up while this option is now depracted as of patch v6.1.9.0 - 17/06/2023
* FIXED a bug when "UPDATE" choice in Illegal Services x IS.bookmarks.html where not showing an internet error message in case you didn't connected to internet.
* IMPROVED IS Source : Code.
* IMPROVED IS Source : More precise error messages at some places.
- REMOVED Main Menu Credits. Instead it will now start IS website credits page, that way it is easy for me to maintain it up-to-date on both IS Windows application and IS website.

v6.1.9.1 - 17/06/2023
* IMPROVED IS Source : Code.

v6.1.9.0 - 17/06/2023
* FIXED a bug blocking people to use IS because it could not work with the "my documents" user's folder. (probably due to OneDrive or others cloud file saving plateforms.)
The fameous: "Illegal Services cannot start because it could not determine its output directory." that a lot of people have reported to me.
Now the "C:\Users\username\Documents\llegal Services" folder changed location and should have automatically moved all files to "C:\Users\username\AppData\Roaming\Illegal Services\user_data\".
* FIXED a potential bug due to a compiler's bug, that was setting the wrong paths at some places in IS.

v6.1.8.9 - 18/03/2023
+ UPDATED IS lib : bookmarks_parser.exe

v6.1.8.8 - 16/03/2023
* FIXED IS Main Menu, Illegal Services x IS.bookmarks.html "IP Denial of Services (DDoS)" displaying "Bookmarks Toolbar/Illegal Services/More Features/Warez Wikis/IP DENIAL OF SERVICES (DDOS)" instead of "Bookmarks Toolbar/Illegal Services/IP Denial of Services (DDoS)".
* FIXED Illegal Services x IS.bookmarks.html case sensivity searches.

v6.1.8.7 - 25/02/2023
* HOTFIX for Quick Batch File Compiler not registered.

v6.1.8.6 - 25/02/2023
- REMOVED IS Proxy : https://git.teknik.io/Illegal-Services/Illegal_Services (down)
* UPDATED IP Denial of Services (DDoS)

v6.1.8.5 - 23/11/2022
* FIXED IS Setup : Support for Windows 11.
* FIXED IS Launcher : Searching a proxy.
* FIXED IS Source : Minor bugs.
* IMPROVED IS Source : Support for Windows 11.
* IMPROVED IS Source : Moved "Portable_Apps", "YouTube DL" and "IS.bookmarks.html" folders and files to the "Illegal Services" user documents folder.
* IMPROVED IS Source : Moved both "Portable_Apps" and "YouTube DL" folders to "Illegal Services" user documents.
* IMPROVED IS Source : Error Messages.
+ ADDED Settings : Scan down or changed domain websites whitelist.
+ ADDED IS Source : more error messages.
+ ADDED IS Source : Now logs are created in the "Illegal Services" user documents under the "logs" folder.
+ ADDED IS Source : Now before downloading IS.bookmarks.html, it checks that the online hash is different from the local one.
+ ADDED More Features : View Hardware Identification.
+ ADDED Settings : Extract IS source code (publicly)
+ UPDATED IS lib : EULA.rtf

v6.0.0.0 - 17/02/2022
* FIXED IS Source : Minor bugs.
* FIXED IS Source : A bug when Illegal Services executable have been renamed with invalid characters.
* FIXED IS Source : A bug when Illegal Services call PowerShell, the screen appears to change size. (thanks @exro)
* FIXED IS Source : A bug when user edits "IP Lookup Saved.txt" but forget the ending newline. (thanks @Grub4K)
* IMPROVED IS Source : Code.
* IMPROVED IS Source : Curl PATH algorithm. (thanks @Grub4K)
* IMPROVED Launcher : Checking files integrity.
* IMPROVED Launcher : Checking Windows version.
+ UPDATED IS lib : curl.exe v7.81.0
+ UPDATED Main Menu Credits.
+ UPDATED Main Menu Our Social Networks.
+ UPDATED Main Menu FAQ.
+ UPDATED IP Address Lookup. (thanks @Rosalyn)
+ ADDED IS lib : binread (thanks @anic17 and @Grub4K)
+ ADDED IS lib : bookmarks_parser.exe (thanks @Grub4K)
+ ADDED Setup : Checking Windows version.
+ ADDED IS Source : IS bookmarks parser. (Illegal Services x IS.bookmarks.html)
+ ADDED IS Source : Support for Windows 11. (thanks @Grub4K and @sintrode)
+ ADDED IS Source : IS.Bookmarks.html parser (thanks @Grub4K, @sintrode and @anic17)
+ ADDED IS Source : Multiple checkbox choices.
+ ADDED More Features : Spoofing. (thanks @0x00)
- REMOVED IS Source : Support for Windows 7, 8, 8.1 (https://github.com/Illegal-Services/Illegal_Services/issues/7)
- REMOVED IS lib : Ansicon

v5.9.0.0 - 19/09/2021
- REDUCED Code. (~12 ko)
* IMPROVED code.
* IMPROVED flagging of untrusted website.
* FIXED Settings : Bugs related to user input in Regedit. (thanks @anic17 and @sintrode)
* FIXED Minimum version of illegal services is now set to Windows 7.
* FIXED a bug with antivirus conflicting with the updater.
* FIXED a bug with an infinite build update loop when "Illegal_Services.exe" does not have the correct build number assigned.
* FIXED a bug when HTTPS was always chosen to start websites.
* FIXED a bug with YouTube DL "change output location" not changing "download location" strings.
* FIXED a bug when the "&" character was entered in the URL checker.
* FIXED minor bugs.
* MOVED Useful Websites : https://doxbin.org/ in Doxing.
* MOVED Useful Websites : https://freezer.life/ in More Features Musics.
+ UPDATED IS lib : curl.exe v7.79.0
+ UPDATED IS lib : EULA.rtf
+ UPDATED IS lib : Tutorial.html (thanks @anic17)
+ UPDATED Main Menu Credits.
+ UPDATED Internet Protocol Television : http://pro.allkaicerteam.com/

Website_Builder.py
import re
import shutil
import textwrap
import datetime
import subprocess
from pathlib import Path
from urllib.parse import urlparse


def encode_unicode_encoding(string: str, type: str):
    if type == "path":
        replacements = {
            '\\\\': 'U+005C',
            '\\/': 'U+002F',
        }
    else:
        replacements = {
            '\\': 'U+005C',
            '/': 'U+002F',
        }
    replacements.update({
        ':': 'U+003A',
        '*': 'U+002A',
        '?': 'U+003F',
        '"': 'U+0022',
        '<': 'U+003C',
        '>': 'U+003E',
        '|': 'U+007C',
    })
    for chars, replacement in replacements.items():
        string = string.replace(chars, replacement)
    return string

def decode_unicode_encoding(string: str, type: str):
    if type == "path":
        replacements = {
            'U+005C': '\\\\',
            'U+002F': '\\/',
        }
    else:
        replacements = {
            'U+005C': '\\',
            'U+002F': '/',
        }
    replacements.update({
        'U+003A': ':',
        'U+002A': '*',
        'U+003F': '?',
        'U+0022': '"',
        'U+003C': '<',
        'U+003E': '>',
        'U+007C': '|',
    })
    for chars, replacement in replacements.items():
        string = string.replace(chars, replacement)
    return string

def encode_html_entity_encoding(string: str):
    replacements = {
        '&': '&amp;',
        '"': '&quot;',
        '\'': '&#39;',
        '<': '&lt;',
        '>': '&gt;',
        ' ': '&nbsp;',
    }
    for chars, replacement in replacements.items():
        string = string.replace(chars, replacement)
    return string

def decode_html_entity_encoding(string: str):
    replacements = {
        '&amp;': '&',
        '&quot;': '"',
        '&#39;': '\'',
        '&lt;': '<',
        '&gt;': '>',
    }
    for chars, replacement in replacements.items():
        string = string.replace(chars, replacement)
    return string

def encode_url_encoding(string: str):
    replacements = {
        '%': '%25',
        ' ': '%20',
        '[': '%5B',
        ']': '%5D',
        '{': '%7B',
        '}': '%7D',
        '^': '%5E',
        '`': '%60',
        '#': '%23',
    }
    for chars, replacement in replacements.items():
        string = string.replace(chars, replacement)
    return string

def decode_url_encoding(string: str):
    replacements = {
        '%25': '%',
        '%20': ' ',
        '%5B': '[',
        '%5D': ']',
        '%7B': '{',
        '%7D': '}',
        '%5E': '^',
        '%60': '`',
        '%23': '#',
    }
    for chars, replacement in replacements.items():
        string = string.replace(chars, replacement)
    return string

def write_html_header() -> None:
    display_pathbar = _folder_path = ""
    for folder in bookmark_path__html_href_path.split("/"):
        if _folder_path:
            _folder_path += f"/{folder}"
        else:
            _folder_path = folder
        html_text = encode_html_entity_encoding(decode_url_encoding(decode_unicode_encoding(folder, "folder")))
        display_pathbar += f'<a href="/Illegal_Services/{_folder_path}/index.html">{html_text}</a> > '
    display_pathbar += "index.html"
    with open(bookmark_path__windows_index_path__str, "a+", encoding="utf-8", newline="\r\n") as file:
        text = f"""
            <!DOCTYPE html>
            <html lang="en-US">

            <head>
                <meta charset="UTF-8">
                <meta name="viewport" content="width=device-width, initial-scale=1">
                <meta name="description" content="Illegal Services Bookmarks">
                <meta name="keywords" content="Illegal, Services, Bookmarks, website">
                <meta name="author" content="IB_U_Z_Z_A_R_Dl">
                <title>IS Bookmarks - Illegal Services</title>
                <link rel="shortcut icon" href="/Illegal_Services/icons/favicon.ico" type="image/x-icon">
                <link rel="icon" href="/Illegal_Services/icons/favicon.ico" type="image/x-icon">
                <link rel="stylesheet" href="/Illegal_Services/css/styles.css">
                <link rel="stylesheet" href="/Illegal_Services/font-awesome-4.7.0/css/font-awesome.min.css">
            </head>

            <body>
                <nav>
                    <div class="navbar">
                        <a href="/Illegal_Services/index.html"><i class="fa fa-home"></i><span class="navbar-text-padding">Home</span></a>
                        <a href="/Illegal_Services/credits.html"><i class="fa fa-handshake-o"></i><span class="navbar-text-padding">Credits</span></a>
                        <a href="/Illegal_Services/tutorial.html"><i class="fa fa-life-ring"></i><span class="navbar-text-padding">Tutorial</span></a>
                        <a href="/Illegal_Services/faq.html"><i class="fa fa-question-circle"></i><span class="navbar-text-padding">FAQ</span></a>
                        <a href="/Illegal_Services/downloads.html"><i class="fa fa-cloud-download"></i><span class="navbar-text-padding">Downloads</span></a>
                        <a href="/Illegal_Services/Bookmarks%20Toolbar/Illegal%20Services/index.html" class="navbar-active"><i class="fa fa-bookmark-o"></i><span class="navbar-text-padding"><u>IS Bookmarks</u></span></a>
                    </div>
                </nav>

                <div class="search-or-request_parent">
                    <div class="search-or-request_container">
                        <div class="search-or-request_child1">
                            <form action="https://eofo904128t6ry8.m.pipedream.net/" method="post" target="_blank">
                                <h4>
                                    <label for="search_link">Search a link or folder in IS database:</label>
                                    <br>
                                    <br>
                                    <input type="text" name="search_link" id="search_link" placeholder="https://example.com/">
                                    <input type="submit" value="Search">
                                </h4>
                            </form>
                        </div>
                        <div class="search-or-request_child2">
                            <form action="https://eolu77sciiyyni6.m.pipedream.net/" method="post" target="_blank">
                                <h4>
                                    <label for="request_link">Request a link to be added in IS database:</label>
                                    <br>
                                    <br>
                                    <input type="text" name="request_link" id="request_link" placeholder="https://example.com/">
                                    <input type="submit" value="Request">
                                </h4>
                            </form>
                        </div>
                    </div>
                </div>

                <div class="pathbar">
                    {display_pathbar}
                </div>
                <br>

                <div class="vertical-menu">
        """
        text = textwrap.dedent(text).removeprefix("\n")
        file.write(text)
        file.close()
    if not (bookmark_path__windows_index_path__path.exists() and bookmark_path__windows_index_path__path.is_file()):
        print(f'ERROR (write_html_header): "{bookmark_path__windows_index_path__path}"')
        input()
        exit(0)
    if not (bookmark_path__windows_path__path.exists() and bookmark_path__windows_path__path.is_dir()):
        print(f'ERROR (write_html_header): "{bookmark_path__windows_path__path}"')
        input()
        exit(0)

def write_footer() -> None:
    with open(windows_href_path__str, "a+", encoding="utf-8", newline="\r\n") as file:
        text = """
                </div>
                <br>

                <script src="/Illegal_Services/js/counter.js"></script>
                <noscript>
                    <div class="javascript-disabled">
                        <img src="/Illegal_Services/icons/no_js.png" alt="no_js.png">
                        JavaScript disabled in your browser;<br>
                        can't display the counter informations.
                    </div>
                </noscript>

                <footer>
                    <a href="https://illegal-services.github.io/Illegal_Services/" target="_blank"><img src="/Illegal_Services/svgs/internet.svg" alt="Website" title="https://illegal-services.github.io/Illegal_Services/"></a>
                    <a href="https://github.com/Illegal-Services/Illegal_Services" target="_blank"><img src="/Illegal_Services/svgs/github.svg" alt="GitHub" title="https://github.com/Illegal-Services/Illegal_Services"></a>
                    <a href="https://t.me/illegal_services_forum" target="_blank"><img src="/Illegal_Services/svgs/telegram.svg" alt="Telegram forum" title="https://t.me/illegal_services_forum"></a>
                    <a href="https://t.me/illegal_services" target="_blank"><img src="/Illegal_Services/svgs/telegram.svg" alt="Telegram channel" title="https://t.me/illegal_services"></a>
                    <br>
                    © 2020-2023 IB_U_Z_Z_A_R_Dl. All rights reserved.
                </footer>
            </body>

            </html>
        """
        text = textwrap.dedent(text).removeprefix("\n")
        file.write(text)
        file.close()
    if not (windows_href_path__path.exists() and windows_href_path__path.is_file()):
        print(f'ERROR (write_footer): "{bookmark_folder__href}" "{bookmark_folder__text}"')
        input()
        exit(0)

def create_folder_or_path(folder_or_path: Path):
    if not (folder_or_path.exists() and folder_or_path.is_dir()):
        folder_or_path.mkdir(parents=False, exist_ok=False)
    if not (folder_or_path.exists() and folder_or_path.is_dir()):
        print(f'ERROR (create_folder_or_path): "{folder_or_path}"')
        input()
        exit(0)
    return folder_or_path


links_counter = 0

for file_or_folder in [Path(R"Bookmarks Toolbar"), Path(R"js")]:
    if file_or_folder.exists():
        if file_or_folder.is_dir():
            shutil.rmtree(file_or_folder)
        elif file_or_folder.is_file():
            file_or_folder.unlink()
    if file_or_folder.exists():
        input()
        exit(0)

bookmarks_db = subprocess.check_output([
    R"D:\Git\Illegal_Services\bookmarks_parser.exe",
    "--extended-parsing",
    "--folders-path",
    "--quoting-style",
    "'",
    R"D:\Git\Illegal_Services\IS.bookmarks.html"
]).decode().splitlines(keepends=False)

is_first_folder__flag = False
for bookmark in bookmarks_db:
    parts = re.findall(r"'(.*?)'", bookmark)

    bookmark_type = str(parts[0])
    bookmark_depth = str(parts[1])
    bookmark_path = str(parts[2])

    #print(bookmark)
    #print(parts)

    if (
        bookmark_type == "PATH"
        and bookmark_depth == "0"
        and bookmark_path == ""
    ):
        is_first_folder__flag = True
        bookmark_path = str(parts[3])
    bookmark_path__windows_path__str = encode_unicode_encoding(decode_html_entity_encoding(bookmark_path), "path")
    bookmark_path__windows_path__path = Path(bookmark_path__windows_path__str)
    bookmark_path__windows_index_path__str = f"{bookmark_path__windows_path__str}/index.html"
    bookmark_path__windows_index_path__path = Path(bookmark_path__windows_index_path__str)
    bookmark_path__html_href_path = encode_url_encoding(decode_url_encoding(bookmark_path__windows_path__str))
    bookmark_path__html_href_text = encode_html_entity_encoding(bookmark_path__windows_path__str)
    if not (bookmark_path__windows_path__path.exists() and bookmark_path__windows_path__path.is_dir()):
        create_folder_or_path(bookmark_path__windows_path__path)
        write_html_header()
    if is_first_folder__flag is True:
        is_first_folder__flag = None
        continue

    if bookmark_type == "LINK":
        bookmark_link = str(parts[3])
        bookmark_link_title = str(parts[4])
        links_counter += 1
        bookmark_link_hostname = urlparse(bookmark_link).netloc
        bookmark_link_title__text = encode_html_entity_encoding(decode_html_entity_encoding(bookmark_link_title))

        match = re.search(r"^(.*)( \| \(untrusted(?:\: .*))$", bookmark_link_title)
        if match:
            bookmark_link_title__html_text = f'{match.group(1)}<span style="color: red">{match.group(2)}</span>'
        else:
            bookmark_link_title__html_text = bookmark_link_title__text

        with open(bookmark_path__windows_index_path__str, "a+", encoding="utf-8", newline="\r\n") as file:
            file.write(f'        <a href="{bookmark_link}" target="_blank" title="{bookmark_link}"><img src="https://external-content.duckduckgo.com/ip3/{bookmark_link_hostname}.ico" alt="favicon">{bookmark_link_title__html_text}</a>\n')
            file.close()
        if not (bookmark_path__windows_index_path__path.exists() and bookmark_path__windows_index_path__path.is_file()):
            print(f'ERROR (WRITE_LINK_INDEX): "{bookmark_link}" "{bookmark_folder__text}" "{bookmark_link_title__text}"')
            input()
            exit(0)

    elif bookmark_type == "PATH":
        bookmark_folder = str(parts[3])
        bookmark_folder__href = f"{encode_url_encoding(encode_unicode_encoding(decode_html_entity_encoding(bookmark_folder), 'folder'))}/index.html"
        bookmark_folder__text = encode_html_entity_encoding(decode_html_entity_encoding(bookmark_folder))
        with open(bookmark_path__windows_index_path__str, "a+", encoding="utf-8", newline="\r\n") as file:
            file.write(f'        <a href="{bookmark_folder__href}"><i class="fa fa-folder-o"></i>{bookmark_folder__text}</a>\n')
            file.close()
        if not (bookmark_path__windows_index_path__path.exists() and bookmark_path__windows_index_path__path.is_file()):
            print(f'ERROR (WRITE_PATH_INDEX): "{bookmark_folder__href}" "{bookmark_folder__text}"')
            input()
            exit(0)

    elif bookmark_type == "HR":
        #bookmark_hr = str(parts[3]) # this code, not correctly implemented in "bookmarks_parser.py" (expected: "--------------------" instead of --------------------)
        with open(bookmark_path__windows_index_path__str, "a+", encoding="utf-8", newline="\r\n") as file:
            file.write("        <HR>\n")
            file.close()
        if not (bookmark_path__windows_index_path__path.exists() and bookmark_path__windows_index_path__path.is_file()):
            print(f'ERROR (WRITE_HR_INDEX): "{bookmark_path__windows_index_path__str}"')
            input()
            exit(0)
    else:
        print(True)
        input()
        exit(0)

folder_path = Path("Bookmarks Toolbar")
for file_path in folder_path.glob("**/*.html"):
    if not (file_path.exists() and file_path.is_file()):
        continue

    windows_href_path__path = file_path
    windows_href_path__str = str(file_path)
    if windows_href_path__str:
        write_footer()

js_conter__path = Path("js\counter.js")
create_folder_or_path(js_conter__path.parent.resolve())
with open("js\counter.js", "w", encoding="utf-8") as file:
    text = f"""
        document.write(`
            <div class="counter">
                Updated: {datetime.date.today().strftime("%d/%m/%Y")}&nbsp;&nbsp;|&nbsp;&nbsp;{links_counter} links indexed.
            </div>
            <br>
        `);
    """
    text = textwrap.dedent(text).removeprefix("\n")
    file.write(text)
    file.close()
if not (js_conter__path.exists() and js_conter__path.is_file()):
    print(f'ERROR (write_js_conter): "{bookmark_path__windows_index_path__path}"')
    input()
    exit(0)

exit(0)